### PR TITLE
Feature/aql enumeratecollection executor

### PR DESCRIPTION
**Pull request by hkernbach**
(Also enterprise branch needed)

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/2837/

**Comment by hkernbach** [comment]
Checklilst: 

- [ ] Remove old Blocks
- [ ] Tests green
- [ ] Review

**Comment by hkernbach** [comment]
http://jenkins01:8080/job/arangodb-matrix-pr-linux/4065/

File: arangod/Aql/DocumentProducingHelper.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/OutputAqlItemRow.h"
+#include "Basics/Common.h"
+
+#include <velocypack/Slice.h>
+
+namespace arangodb {
+namespace aql {
+
+class DocumentProducingHelper {
```
Comment (ObiWahn) [inline arangod/Aql/DocumentProducingHelper.h]: Why is a single static function in a class?

File: arangod/Aql/DocumentProducingHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        } else {
+          b.add(it, found);
+        }
+      }
+    }
+  }
+}
+
+// TODO: Importable typdef from header file?
+typedef std::function<void(InputAqlItemRow&, OutputAqlItemRow&, arangodb::velocypack::Slice, RegisterId)> DocumentProducingFunction;
```
Comment (ObiWahn) [inline arangod/Aql/DocumentProducingHelper.cpp]: Use type defined in the header.

File: arangod/Aql/ExecutionNode.cpp [hunk 3]
```diff
@@ -1256,7 +1258,27 @@ void EnumerateCollectionNode::toVelocyPackHelper(VPackBuilder& builder, unsigned
...
+  auto it = getRegisterPlan()->varInfo.find(_outVariable->id);
+  TRI_ASSERT(it != getRegisterPlan()->varInfo.end());
+  RegisterId outputRegister = it->second.registerId;
+
+  // Variable const* outVariable = _outVariable;
+  // outVariable = _plan->getAst()->variables()->createVariable(outVariable);
+  // TRI_ASSERT(outVariable != nullptr);
+
+  transaction::Methods* trxPtr = _plan->getAst()->query()->trx();
+  bool allowCoveringIndexOptimization = true;
```
Comment (ObiWahn) [inline arangod/Aql/ExecutionNode.cpp]: Is this always true?

Comment (hkernbach) [reply arangod/Aql/ExecutionNode.cpp]: In this case yes, but another future class is setting another parameter. This is currently as it is for better readability (and will anyway be optimized out?). If not I'll exchange this with a comment.

File: arangod/Aql/DocumentProducingHelper.h (hunk 1 above)
Comment (hkernbach) [reply arangod/Aql/DocumentProducingHelper.h]: To be extendable in the future? What would be a better approach?

File: arangod/Aql/ExecutionNode.cpp (hunk 3 above)
Comment (ObiWahn) [reply arangod/Aql/ExecutionNode.cpp]: No the variable is totally fine as long as it is set to false at some point (e.g. index enumerator)

File: arangod/Aql/DocumentProducingHelper.h (hunk 1 above)
Comment (ObiWahn) [reply arangod/Aql/DocumentProducingHelper.h]: I am not sure in which way it needs to be extendable but, to me it really looks like the static function should rather be a free-standing-function.

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/2868/

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/2877/

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/2898/

**Pull request merged**